src/t_zset.c

```diff
@@ -335,34 +335,46 @@ int zslIsInRange(zskiplist *zsl, zrangespec *range) {
 zskiplistNode *zslNthInRange(zskiplist *zsl, zrangespec *range, long n) {
     zskiplistNode *x;
     int i;
-    /* rank is 1-based. */
-    long rank = 0;
+    long edge_rank = 0;
+    long last_highest_level_rank = 0;
+    zskiplistNode *last_highest_level_node = NULL;
+    unsigned long rank_diff;
 
     /* If everything is out of range, return early. */
     if (!zslIsInRange(zsl,range)) return NULL;
 
+    /* Go forward while *OUT* of range at level of zsl->level-1. */
     x = zsl->header;
+    i = zsl->level - 1;
+    while (x->level[i].forward && !zslValueGteMin(x->level[i].forward->score, range)) {
+        edge_rank += x->level[i].span;
+        x = x->level[i].forward;
+    }
+    /* Remember the last node which has zsl->level-1 levels and its rank. */
+    last_highest_level_node = x;
+    last_highest_level_rank = edge_rank;
+
     if (n >= 0) {
-        for (i = zsl->level - 1; i >= 0; i--) {
+        for (i = zsl->level - 2; i >= 0; i--) {
             /* Go forward while *OUT* of range. */
             while (x->level[i].forward && !zslValueGteMin(x->level[i].forward->score, range)) {
                 /* Count the rank of the last element smaller than the range. */
-                rank += x->level[i].span;
+                edge_rank += x->level[i].span;
                 x = x->level[i].forward;
             }
         }
         /* Check if zsl is long enough. */
-        if ((unsigned long)(rank + n) >= zsl->length) return NULL; 
+        if ((unsigned long)(edge_rank + n) >= zsl->length) return NULL;
         if (n < ZSKIPLIST_MAX_SEARCH) {
             /* If offset is small, we can just jump node by node */
             /* rank+1 is the first element in range, so we need n+1 steps to reach target. */
             for (i = 0; i < n + 1; i++) { 
                 x = x->level[0].forward;
             }
         } else {
-            /* If offset is big, we jump to the nth node by its rank from the start of skiplist. */
-            /* rank+1 is the first element in range, and n is 0-based, so we need rank+n+1. */
-            x = zslGetElementByRank(zsl, rank + n + 1);
+            /* If offset is big, we can jump from the last zsl->level-1 node. */
+            rank_diff = edge_rank + 1 + n - last_highest_level_rank;
+            x = zslGetElementByRank(last_highest_level_node, zsl->level - 1, rank_diff);
```

Comment:
so now that we have this, do we really need ZSKIPLIST_MAX_SEARCH and the above `if-else` branch?

Reply:
This is still useful when offset is reversed and range is very big, so that we don't need to jump across a big range.

Reply:
btw, i wonder again if we need it for the forward search, or that maybe we CAN somehow use multiple levels in the forward search, without falling back to the last `[zsl->level - 1]` node.
i.e. 
![image](https://github.com/redis/redis/assets/7045099/a8df1f86-4cb8-4a58-8fd3-abdef89eeb4d)

for that, we need a way to know how many levels we have in each `level[]` array.
we can take one bit from the `unsigned long span` field to indicate if it's the last element in the `level[]` array.
and then we can do the blue walk in the above diagram.
WDYT?

Reply:
yes, i agree it adds complexity.
my thinking was that it'll remove the need for the ZSKIPLIST_MAX_SEARCH threshold and simplify zslGetElementByRankFromNode (you won't need to pass the level hint to it).

but since that's not gonna solve the backwards walk problem, maybe we shouldn't bother